crates/engine/src/rocksdb_engine/transaction.rs

```diff
@@ -1,18 +1,395 @@
 #![allow(clippy::module_name_repetitions)]
 #![allow(clippy::multiple_inherent_impl)]
 
-use crate::TransactionApi;
+use std::{
+    iter::repeat,
+    sync::{atomic::AtomicU64, Arc},
+};
 
-/// A transaction of the `RocksEngine`
-#[derive(Copy, Clone, Debug, Default)]
-pub struct RocksTransaction;
+use clippy_utilities::{NumericCast, OverflowArithmetic};
+use parking_lot::Mutex;
+use rocksdb::{
+    Direction, IteratorMode, OptimisticTransactionDB, Transaction, WriteBatchWithTransaction,
+    WriteOptions,
+};
 
+use crate::{
+    api::transaction_api::TransactionApi, error::EngineError, rocksdb_engine::RocksEngine,
+    StorageOps, WriteOperation,
+};
+
+/// Transaction type for `RocksDB`
+#[derive(Debug)]
+pub struct RocksTransaction {
+    /// Inner state
+    inner: Mutex<Option<Inner>>,
+}
+
+/// Inner state of the transaction
+///
+/// WARN: `db` should never be dropped before `txn`
```

Comment:
Is it a gentleman agreement? Do we have any method to ensure that this warning will never happen?

Reply:
Unsafe code removed.